src/dune_pkg/opam_solver.ml

```diff
@@ -256,35 +256,60 @@ let opam_commands_to_actions package (commands : OpamTypes.command list) =
-          (match pform_of_ident_opt ident with
-           | Ok pform -> String_with_vars.make_pform Loc.none pform
-           | Error (`Unknown_variable variable) ->
-             (* Note that the variable name is always quoted to clarify
-                the error message in cases where the grammar of the
-                sentence would otherwise be unclear, such as:
-
-                - Encountered unknown variable type while processing...
-                - Encountered unknown variable name while processing...
-
-                In these examples, the words "type" and "name" are variable
-                names but it would be easy for users to misunderstand those
-                error messages without quotes. *)
+    let pform_of_ident_exn ident =
+      match pform_of_ident ident with
+      | Ok pform -> pform
+      | Error (`Unknown_variable variable) ->
+        (* Note that the variable name is always quoted to clarify
+           the error message in cases where the grammar of the
+           sentence would otherwise be unclear, such as:
+
+           - Encountered unknown variable type while processing...
+           - Encountered unknown variable name while processing...
+
+           In these examples, the words "type" and "name" are variable
+           names but it would be easy for users to misunderstand those
+           error messages without quotes. *)
+        User_error.raise
+          [ Pp.textf
+              "Encountered unknown variable %S while processing commands for package %s."
+              variable
+              (OpamPackage.to_string package)
+          ; Pp.text "The full command:"
+          ; Pp.text (opam_command_to_string_debug command)
+          ]
+    in
+    let interpolate_opam_variables s =
+      Dune_re.Seq.split_full OpamFilter.string_interp_regex s
+      |> Seq.map ~f:(function
+        | `Text text -> `Text text
+        | `Delim group ->
+          (match Dune_re.Group.get group 0 with
+           | "%%" -> `Text "%"
+           | interp
+             when String.starts_with ~prefix:"%{" interp
+                  && String.ends_with ~suffix:"}%" interp ->
+             let ident = String.sub ~pos:2 ~len:(String.length interp - 4) interp in
+             `Pform (pform_of_ident_exn ident)
+           | other ->
              User_error.raise
                [ Pp.textf
-                   "Encountered unknown variable %S while processing commands for \
-                    package %s."
-                   variable
+                   "Encountered malformed variable interpolation while processing \
+                    commands for package %s."
                    (OpamPackage.to_string package)
+               ; Pp.text "The variable interpolation:"
+               ; Pp.text other
                ; Pp.text "The full command:"
                ; Pp.text (opam_command_to_string_debug command)
                ]))
+      |> List.of_seq
+      |> String_with_vars.make Loc.none
```

Comment:
This creates the `String_with_vars.t` of out the split result and implicitly adds `" "` as a separator, but what I don't understand why this would be necessary — I don't see where the spaces are getting consumed so they would need to be added. Can you clarify?

Reply:
The default separator is `""`, not `" "`. I'll look into simplifying the api as per your other suggestions and also make the separator explicit to clarify cases like this.

Reply:
Actually turns out we don't need the separator anymore so I'll remove it altogether.